typing/typetexp.ml

```diff
@@ -455,7 +475,15 @@ and transl_type_aux env policy mode styp =
         | (l, arg_mode, arg) :: rest ->
           check_arg_type arg;
           let arg_cty = transl_type env policy arg_mode arg in
-          let acc_mode = Alloc_mode.join_const acc_mode arg_mode in
+          let acc_mode =
+            Alloc.Const.join
+              (Alloc.Const.uncurried_ret_mode_from_arg arg_mode)
+              (Alloc.Const.uncurried_ret_mode_from_alloc acc_mode)
+          in
+          let acc_mode =
+            Alloc.Const.join acc_mode
+              (Alloc.Const.min_with_uniqueness Uniqueness.Const.Shared)
```

Comment:
Is this needed? Isn't `uncurried_ret_mode_from_alloc` always `Shared` so the join should also already be `Shared`?

Reply:
Both `close_over` and `partial_apply` give `unique` because I think they are used as lower bound, and we don't want to constrain the uniqueness axis of functions. When interpreting type signatures, however, we want functions always to be `shared`. That means `Mode.close_over` and `Mode.Const.close_over` would have different behaviours on the uniqueness axis. I feel I should avoid that, so the two have the same behaviours, but the caller will do some extra tweaking.